### PR TITLE
Change: compile markdown at data hook

### LINKED ISSUE #143: [markdown] TypeError when the source has no front matter
```
C:\Users\eight04\dev\fancy-t-spin\node_modules\@elderjs\plugin-markdown\index.js:132
            if (md.frontmatter.draft || md.slug.indexOf('draft') === 0) {
                               ^

TypeError: Cannot read properties of undefined (reading 'draft')
    at C:\Users\eight04\dev\fancy-t-spin\node_modules\@elderjs\plugin-markdown\index.js:132:32
    at Array.forEach (<anonymous>)
    at C:\Users\eight04\dev\fancy-t-spin\node_modules\@elderjs\plugin-markdown\index.js:130:34
    at Array.forEach (<anonymous>)
    at Object.init (C:\Users\eight04\dev\fancy-t-spin\node_modules\@elderjs\plugin-markdown\index.js:124:36)
    at async Object.plugins [as default] (C:\Users\eight04\dev\fancy-t-spin\node_modules\@elderjs\elderjs\build\plugins\index.js:78:18)

Node.js v17.4.0
```
https://github.com/Elderjs/plugins/blob/9626156b4464b2dd1261f428c9e47281a69395bb/packages/markdown/index.js#L85
Should probably change to:
```js
data: { frontmatter = {}, ...data }
```

### LINKED ISSUE #144: [markdown] Process markdown in data hook?
Currently, markdown is compiled in `init()`, and it seems that a large markdown file can block the entire process when developing (`npm run dev`).

Will it be better if we move compilation into a data hook? And maybe move it to a subprocess? As I understood, data hook is only called when requested.

I think we can prepare slug and front matter in `init()` and move everything else to data hook.

### PR DESCRIPTION
**Pull request by eight04**
Fixes #143
Fixes #144

Also fixed an issue that the first argument of `slugFormatter` (`file`) is not a relative file path on Windows.

### CODE REVIEW

File: packages/markdown/index.js [hunk 1]
```diff
@@ -65,49 +64,19 @@ const plugin = {
...
-            data,
+          const markdown = createMarkdownStore({
+            root: mdsInRoute,
+            file,
+            parser: plugin.markdownParser,
+            useImagePlugin: plugin.settings.plugins['@elderjs/plugin-images'] && plugin.config.useElderJsPluginImages,
+            shortcodes: plugin.settings.shortcodes,
+            slug: plugin.config.slugFormatter,
           });
+          await markdown.prepareSlug();
```
Comment (eight04) [inline packages/markdown/index.js:75]: TBH I don't know if this approach does improve performance.

To benchmark, we can change this line to `await markdown.prepareHtml()` to use the old behavior.

Comment (nickreese) [reply packages/markdown/index.js:75]: I believe this should be quiet a bit faster due to not processing all of the html. Also should help out of memory issues on huge sites.

**Comment by nickreese** [comment]
Hey again @eight04, thanks for the contribution.

I like this pattern but overall feel the usage is a bit confusing and a quick refactor would make it easier to logic through. 

In my mental model it makes more sense to call `await createMarkdownStore({...})` which in turn is async and invokes both `prepareSlug`, `prepareFrontMatter`, and `prepareSource` all on the bootstrap hook.

Yes bootstrap is can get slow on dev, but the design is such that was much preprocessing is done ahead of time so it should be fast when in SSR mode. 

I like the test coverage and the two fixes and this should result in a major speed increase for dev if we aren't parsing and preprocessing all of the html ahead of time.

**Comment by nickreese** [comment]
You'll also see that I bumped up the `remark-html` just to get the critical vuln (which doesn't apply to use) out of the `npm i`

**Comment by eight04** [comment]
When testing, my build time does reduced like 2 seconds. I think it is because that the build splits into 4 processes on my machine therefore a single markdown used to be read/compiled 4 times.

If the build only uses a single process, this change would probably decrease the performance. 

Still improves the dev though.

**Comment by eight04** [comment]
Done.

**Pull request merged**